File: solar_scrapping/solax/solax.py
import os
import time
import json
import pandas as pd
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# Configuration
LOGIN_URL = "https://www.solaxcloud.com/user-center/"
USERNAME = "oaksun"
PASSWORD = "Oaksun@1006@"
EXCEL_FILE = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "solar_data.xlsx"))

# Set to True if running in background without a screen
HEADLESS = True

PLANT_MAPPING = {
    "my space study hall 1": 1,
    "my space study hall": 2,
    "vasudev blue flied site": 3,
    "villa num 111": 4,
    "rv2": 5,
    "rv1": 6,
    "melgiri enterprises house 3": 7,
    "melgiri enterprises": 8,
    "melgiri enterprises house 2": 9,
    "melgiri farm 02": 10,
    "melgiri farms 01": 11,
    "magna villa 15": 12,
    "abhilash reddy incrible hallmark villa no.24": 13,
    "arya vysya anna satram": 14,
    "central university of haryana": 15,
    "magna villa 41": 16
}

def get_plant_id(plant_name):
    key = str(plant_name).strip().lower()
    if key in PLANT_MAPPING:
        return PLANT_MAPPING[key]
    new_id = max(PLANT_MAPPING.values()) + 1 if PLANT_MAPPING else 1
    PLANT_MAPPING[key] = new_id
    print(f"Warning: Dynamic plant mapping created for '{plant_name}' -> ID {new_id}")
    return new_id

def clean_timestamp(date_str):
    if pd.isna(date_str) or not date_str:
        return None
    s = str(date_str).strip()
    s = s.split('(')[0].strip()
    
    parts = s.split(' ')
    if len(parts) >= 2:
        date_part = parts[0]
        time_part = parts[1]
        for sep in ['/', '-']:
            if sep in date_part:
                date_pieces = date_part.split(sep)
                if len(date_pieces) == 3:
                    if len(date_pieces[0]) == 4:
                        return f"{date_pieces[0]}-{date_pieces[1].zfill(2)}-{date_pieces[2].zfill(2)} {time_part}"
                    else:
                        day = date_pieces[0].zfill(2)
                        month = date_pieces[1].zfill(2)
                        year = date_pieces[2]
                        return f"{year}-{month}-{day} {time_part}"
    try:
        dt = pd.to_datetime(s)
        return dt.strftime("%Y-%m-%d %H:%M:%S")
    except:
        return s

def parse_value_with_unit(val_str):
    if pd.isna(val_str) or val_str is None:
        return None
    s = str(val_str).strip().lower()
    if s == "" or s == "--":
        return None
    
    multiplier = 1.0
    clean = s
    if s.endswith('mwh'):
        multiplier = 1000.0
        clean = s[:-3].strip()
    elif s.endswith('kwh'):
        clean = s[:-3].strip()
    elif s.endswith('wh'):
        multiplier = 0.001
        clean = s[:-2].strip()
    elif s.endswith('kwp'):
        clean = s[:-3].strip()
    elif s.endswith('kw'):
        clean = s[:-2].strip()
    elif s.endswith('w'):
        multiplier = 0.001
        clean = s[:-1].strip()
    elif s.endswith('h'):
        clean = s[:-1].strip()
    elif s.endswith('°c') or s.endswith('c'):
        clean = s[:-2].strip() if s.endswith('°c') else s[:-1].strip()
        
    try:
        return float(clean) * multiplier
    except:
        try:
            return float(clean)
        except:
            return None

def init_driver():
    print("Initializing Chrome browser...")
    options = webdriver.ChromeOptions()
    options.add_argument("--start-maximized")
    options.add_argument("--disable-gpu")
    
    if HEADLESS:
        print("Running in headless mode...")
        options.add_argument("--headless=new")
        options.add_argument("--window-size=1920,1080")
        options.add_argument("--no-sandbox")
        options.add_argument("--disable-dev-shm-usage")
        options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36")
        
    driver = webdriver.Chrome(options=options)
    return driver

def handle_login(driver):
    print(f"Navigating to login page: {LOGIN_URL}")
    driver.get(LOGIN_URL)
    
    # Wait for input fields to be visible
    wait = WebDriverWait(driver, 15)
    print("Locating credentials input fields...")
    
    # User name field
    username_field = wait.until(
        EC.presence_of_element_located((By.XPATH, "//input[@placeholder='Please enter your user name/email/mobile number']"))
    )
    # Password field
    password_field = wait.until(
        EC.presence_of_element_located((By.XPATH, "//input[@placeholder='Enter Password']"))
    )
    
    # Enter credentials
    print("Entering credentials...")
    username_field.clear()
    username_field.send_keys(USERNAME)
    time.sleep(0.5)
    password_field.clear()
    password_field.send_keys(PASSWORD)
    time.sleep(0.5)
    
    # Find and handle Privacy Agreement checkbox
    print("Checking Privacy Policy agreement checkbox...")
    try:
        # Search for checkbox within the label containing 'Privacy Policy'
        agree_checkbox = driver.find_element(By.XPATH, "//span[contains(text(), 'Privacy Policy') or contains(text(), 'agree to')]/ancestor::label//input[@type='checkbox']")
        if not agree_checkbox.is_selected():
            print("Agreement checkbox not selected, clicking it via JS...")
            driver.execute_script("arguments[0].click();", agree_checkbox)
            time.sleep(0.5)
        else:
            print("Agreement checkbox is already selected.")
    except Exception as e:
        print(f"Warning/Note while checking agreement checkbox: {e}")
        
    # Click Login button
    print("Clicking Login button...")
    login_btn = wait.until(
        EC.element_to_be_clickable((By.XPATH, "//*[contains(@class, 'submit-button') or contains(@class, 'login-btn')]"))
    )
    login_btn.click()
    
    # Wait for dashboard transition
    print("Waiting for page redirect and dashboard load...")
    wait.until(EC.url_contains("/green/#/"))
    print("Login successful!")
    time.sleep(3)

def go_to_plants_page(driver):
    print("Navigating to Plants section...")
    wait = WebDriverWait(driver, 40)
    
    # Find Plants in sidebar and click it
    plants_tab = wait.until(
        EC.element_to_be_clickable((By.XPATH, "//span[text()='Plants']"))
    )
    plants_tab.click()
    
    print("Waiting for Plants table data to load...")
    try:
        # Wait for global progress bar to complete
        wait.until(
            lambda d: "nprogress-busy" not in (d.find_element(By.TAG_NAME, "html").get_attribute("class") or "")
        )
        # Wait for at least one data cell to render
        wait.until(
            EC.presence_of_element_located((By.CLASS_NAME, "arco-table-td"))
        )
        time.sleep(3) # Extra render buffer
    except Exception as e:
        print(f"Warning: Timed out waiting for table to load: {e}")

def scrape_plants_table(driver):
    print("Scraping Plants overview table...")
    
    # Inject JavaScript to extract the table columns and rows dynamically
    js_scrape_table = """
    try {
        const headers = Array.from(document.querySelectorAll('.arco-table-th')).map(th => th.innerText.trim().replace(/\\n/g, ' '));
        const rows = document.querySelectorAll('.arco-table-tr');
        const results = [];
        rows.forEach(row => {
            const cells = Array.from(row.querySelectorAll('.arco-table-td'));
            if (cells.length > 0) {
                const rowData = {};
                cells.forEach((cell, idx) => {
                    const header = headers[idx] || `Column_${idx}`;
                    rowData[header] = cell.innerText.trim().replace(/\\n/g, ' ');
                });
                results.push(rowData);
            }
        });
        return results;
    } catch (e) {
        return [];
    }
    """
    table_data = driver.execute_script(js_scrape_table)
    print(f"Found {len(table_data)} plants in the table.")
    return table_data

def scrape_plant_details(driver):
    print("Scraping plant detailed metrics from dashboard...")
    
    # Inject JavaScript to find values associated with specific metric labels dynamically
    js_scrape_details = """
    try {
        const metrics = {};
        const elements = Array.from(document.querySelectorAll('*'));
        
        const findValue = (label) => {
            const candidates = elements.filter(el => {
                const txt = (el.innerText || "").trim().toLowerCase();
                return txt.includes(label.toLowerCase());
            });
            
            if (candidates.length === 0) return "";
            
            // Sort by children count ascending to get the deepest element containing the label first
            candidates.sort((a, b) => a.getElementsByTagName('*').length - b.getElementsByTagName('*').length);
            
            for (const labelEl of candidates) {
                let current = labelEl;
                for (let i = 0; i < 4; i++) {
                    if (!current) break;
                    const text = current.innerText;
                    const match = text.match(/(\\d+(?:\\.\\d+)?)\\s*(kWh|kWp|kW|%)/i);
                    if (match) {
                        return match[0];
                    }
                    current = current.parentElement;
                }
            }
            return "";
        };
        
        metrics['Daily solar'] = findValue('Daily solar');
        metrics['Daily consumption'] = findValue('Daily consumption');
        metrics['PV Capacity'] = findValue('PV Capacity');
        metrics['PV Power'] = findValue('PV Power');
        metrics['Imported energy'] = findValue('Imported energy today') || findValue('Imported energy');
        
        return metrics;
    } catch (e) {
        return {};
    }
    """
    details = driver.execute_script(js_scrape_details)
    print(f"Scraped details: {details}")
    return details

def save_to_excel(df_new):
    print(f"Saving data to Excel file: {EXCEL_FILE}")
    
    df_existing = pd.DataFrame()
    if os.path.exists(EXCEL_FILE):
        try:
            excel_file = pd.ExcelFile(EXCEL_FILE)
            if "Telemetry" in excel_file.sheet_names:
                df_existing = pd.read_excel(EXCEL_FILE, sheet_name="Telemetry")
        except Exception as e:
            print(f"Warning: Could not read existing Excel file ({e}).")
            
    if not df_existing.empty:
        df_existing['plant_id'] = df_existing['plant_id'].astype(int)
        df_existing['timestamp'] = df_existing['timestamp'].astype(str)
        df_new['plant_id'] = df_new['plant_id'].astype(int)
        df_new['timestamp'] = df_new['timestamp'].astype(str)
        df_combined = pd.concat([df_existing, df_new], ignore_index=True)
    else:
        df_combined = df_new
        
    # Deduplicate rows by plant_id and timestamp
    df_combined = df_combined.sort_values(by="timestamp", ascending=False)
    df_combined = df_combined.drop_duplicates(subset=["plant_id", "timestamp"], keep="first")
    df_combined = df_combined.sort_values(by=["timestamp", "plant_id"]).reset_index(drop=True)
    
    # Re-assign id
    df_combined["id"] = df_combined.index + 1
    
    # Ensure correct columns order
    cols_order = ["id", "plant_id", "timestamp", "power", "voltage", "current", "frequency", 
                  "irradiance", "daily_generation", "total_generation", "temperature", "status", "raw_json", "created_at"]
    for col in cols_order:
        if col not in df_combined.columns:
            df_combined[col] = None
    df_combined = df_combined[cols_order]
    
    try:
        with pd.ExcelWriter(EXCEL_FILE, engine="openpyxl") as writer:
            df_combined.to_excel(writer, sheet_name="Telemetry", index=False)
            
            plants_data = [{"id": pid, "plant_name": name.upper()} for name, pid in PLANT_MAPPING.items()]
            df_plants = pd.DataFrame(plants_data).sort_values(by="id")
            df_plants.to_excel(writer, sheet_name="Plants", index=False)
        print(f"SUCCESS! Unified Telemetry sheet updated. Total {len(df_combined)} rows written.")
    except PermissionError:
        backup = os.path.join(os.path.dirname(EXCEL_FILE), f"solar_data_backup_{int(time.time())}.xlsx")
        print(f"\nERROR: Permission denied writing to '{EXCEL_FILE}'. Saving to backup '{backup}' instead.")
        try:
            with pd.ExcelWriter(backup, engine="openpyxl") as writer:
                df_combined.to_excel(writer, sheet_name="Telemetry", index=False)
        except Exception as e:
            print(f"Could not write backup: {e}")

def main():
    driver = init_driver()
    all_rows = []
    
    try:
        # 1. Log in
        handle_login(driver)
        
        # 2. Navigate to Plants page
        go_to_plants_page(driver)
        
        # 3. Scrape main plants list table
        plants = scrape_plants_table(driver)
        if not plants:
            print("ERROR: No plants data found. Exiting.")
            return
            
        scrape_time = time.strftime("%Y-%m-%d %H:%M:%S")
        
        # 4. Loop through each plant to collect detailed page metrics
        for index, plant in enumerate(plants):
            plant_name = plant.get("Plant Name", "").strip()
            if not plant_name:
                # Handle cases where header might be slightly different
                plant_name = plant.get(list(plant.keys())[0], "").strip()
                
            print(f"\n[{index+1}/{len(plants)}] Processing plant: '{plant_name}'")
            
            # Click on the plant link
            print(f"Navigating to plant details for '{plant_name}'...")
            original_windows = driver.window_handles[:]
            
            # Locate plant link dynamically using text match
            plant_link = WebDriverWait(driver, 15).until(
                EC.element_to_be_clickable((By.XPATH, f"//table//tr//td[contains(., '{plant_name}')]//*[text()='{plant_name}'] | //*[text()='{plant_name}']"))
            )
            
            # Click it
            driver.execute_script("arguments[0].click();", plant_link)
            time.sleep(3) # Initial short sleep for tab to open
            
            # Handle new tab or same tab navigation
            new_windows = driver.window_handles
            opened_in_new_tab = len(new_windows) > len(original_windows)
            
            if opened_in_new_tab:
                new_tab = [w for w in new_windows if w not in original_windows][0]
                driver.switch_to.window(new_tab)
            
            # Wait for details page to render by waiting for nprogress-busy loading to complete
            print("Waiting for plant details page to finish loading...")
            try:
                # Wait for the HTML tag to no longer have class 'nprogress-busy'
                WebDriverWait(driver, 25).until(
                    lambda d: "nprogress-busy" not in (d.find_element(By.TAG_NAME, "html").get_attribute("class") or "")
                )
                # Wait for 'PV Power' label to be present
                WebDriverWait(driver, 15).until(
                    EC.presence_of_element_located((By.XPATH, "//*[text()='PV Power' or contains(text(), 'PV Power')]"))
                )
                time.sleep(4) # Sleep 4 seconds to ensure dynamic numeric values populate
            except Exception as e:
                print(f"Warning: Timed out waiting for details page elements: {e}")
                
            # Scrape detailed metrics
            details = scrape_plant_details(driver)
            
            # Close detail window if opened in new tab, or navigate back
            if opened_in_new_tab:
                driver.close()
                driver.switch_to.window(original_windows[0])
            else:
                print("Navigating back to Plants overview page...")
                driver.get("https://global.solaxcloud.com/green/#/plant/index")
                time.sleep(4)
                
            # Compile merged row data
            row_data = {
                "plant_id": int(get_plant_id(plant_name)),
                "timestamp": str(clean_timestamp(scrape_time)),
                "power": parse_value_with_unit(details.get("PV Power") or plant.get("PV Capacity(kWp)") or plant.get("PV Capacity")),
                "voltage": None,
                "current": None,
                "frequency": None,
                "irradiance": None,
                "daily_generation": parse_value_with_unit(details.get("Daily solar") or plant.get("Daily Yield(kWh)") or plant.get("Daily Yield")),
                "total_generation": None,
                "temperature": None,
                "status": plant.get("Plant Status", "Normal"),
                "raw_json": json.dumps({**plant, **details}, default=str),
                "created_at": time.strftime("%Y-%m-%d %H:%M:%S")
            }
            all_rows.append(row_data)
            
        # 5. Export compiled dataset to Excel
        if all_rows:
            df = pd.DataFrame(all_rows)
            save_to_excel(df)
            
    except Exception as e:
        print(f"\nAn error occurred during scraping: {e}")
        try:
            print(f"Current URL at error: {driver.current_url}")
        except Exception:
            pass
        import traceback
        traceback.print_exc()
        try:
            screenshot_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "error_screenshot.png")
            driver.save_screenshot(screenshot_path)
            print(f"Saved error screenshot to: {screenshot_path}")
        except Exception as se:
            print(f"Could not save screenshot: {se}")
    finally:
        print("\nClosing Chrome browser...")
        driver.quit()

if __name__ == "__main__":
    main()

File: solar_scrapping/solax/solax.js
const puppeteer = require('puppeteer');
// const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Configuration
const LOGIN_URL = "https://www.solaxcloud.com/user-center/";
const JSON_FILE = path.resolve(__dirname, "..", "solar_data.json");
const CREDENTIALS_FILE = path.resolve(__dirname, "..", "credentials.json");

let USERNAME = "oaksun";
let PASSWORD = "Oaksun@1006@";

if (fs.existsSync(CREDENTIALS_FILE)) {
    try {
        const creds = JSON.parse(fs.readFileSync(CREDENTIALS_FILE, 'utf8'));
        if (creds.solax && creds.solax.username && creds.solax.password) {
            USERNAME = creds.solax.username;
            PASSWORD = creds.solax.password;
            console.log("Loaded Solax credentials from credentials.json");
        }
    } catch (e) {
        console.error("Warning: Failed to parse credentials.json, using defaults.", e.message);
    }
}

const PLANT_MAPPING = {
    "my space study hall 1": 1,
    "my space study hall": 2,
    "vasudev blue flied site": 3,
    "villa num 111": 4,
    "rv2": 5,
    "rv1": 6,
    "melgiri enterprises house 3": 7,
    "melgiri enterprises": 8,
    "melgiri enterprises house 2": 9,
    "melgiri farm 02": 10,
    "melgiri farms 01": 11,
    "magna villa 15": 12,
    "abhilash reddy incrible hallmark villa no.24": 13,
    "arya vysya anna satram": 14,
    "central university of haryana": 15,
    "magna villa 41": 16
};

let maxPlantId = 16;
function getPlantId(plantName) {
    const key = String(plantName).trim().toLowerCase();
    if (PLANT_MAPPING[key]) {
        return PLANT_MAPPING[key];
    }
    maxPlantId++;
    PLANT_MAPPING[key] = maxPlantId;
    console.log(`Warning: Dynamic plant mapping created for '${plantName}' -> ID ${maxPlantId}`);
    return maxPlantId;
}

function cleanTimestamp(dateStr) {
    if (!dateStr) return null;
    let s = String(dateStr).trim();
    s = s.split('(')[0].trim();
    
    const parts = s.split(' ');
    if (parts.length >= 2) {
        const datePart = parts[0];
        const timePart = parts[1];
        const separators = ['/', '-'];
        for (const sep of separators) {
            if (datePart.includes(sep)) {
                const datePieces = datePart.split(sep);
                if (datePieces.length === 3) {
                    if (datePieces[0].length === 4) {
                        return `${datePieces[0]}-${datePieces[1].padStart(2, '0')}-${datePieces[2].padStart(2, '0')} ${timePart}`;
                    } else {
                        const day = datePieces[0].padStart(2, '0');
                        const month = datePieces[1].padStart(2, '0');
                        const year = datePieces[2];
                        return `${year}-${month}-${day} ${timePart}`;
                    }
                }
            }
        }
    }
    
    try {
        const d = new Date(s);
        if (!isNaN(d.getTime())) {
            const year = d.getFullYear();
            const month = String(d.getMonth() + 1).padStart(2, '0');
            const date = String(d.getDate()).padStart(2, '0');
            const hours = String(d.getHours()).padStart(2, '0');
            const minutes = String(d.getMinutes()).padStart(2, '0');
            const seconds = String(d.getSeconds()).padStart(2, '0');
            return `${year}-${month}-${date} ${hours}:${minutes}:${seconds}`;
        }
    } catch (e) {}
    
    return s;
}

// Universal parser for numbers with units
function parseValueWithUnit(valStr) {
    if (valStr === undefined || valStr === null) return null;
    const s = String(valStr).trim().toLowerCase();
    if (s === '' || s === '--') return null;
    
    let multiplier = 1;
    let clean = s;
    if (s.endsWith('mwh')) {
        multiplier = 1000;
        clean = s.slice(0, -3).trim();
    } else if (s.endsWith('kwh')) {
        clean = s.slice(0, -3).trim();
    } else if (s.endsWith('wh')) {
        multiplier = 0.001;
        clean = s.slice(0, -2).trim();
    } else if (s.endsWith('kwp')) {
        clean = s.slice(0, -3).trim();
    } else if (s.endsWith('kw')) {
        clean = s.slice(0, -2).trim();
    } else if (s.endsWith('w')) {
        multiplier = 0.001;
        clean = s.slice(0, -1).trim();
    } else if (s.endsWith('h')) {
        clean = s.slice(0, -1).trim();
    } else if (s.endsWith('°c') || s.endsWith('c')) {
        clean = s.endsWith('°c') ? s.slice(0, -2).trim() : s.slice(0, -1).trim();
    }
    const val = parseFloat(clean);
    return isNaN(val) ? null : val * multiplier;
}

// Set to true to run in headless mode
const HEADLESS = true;

const delay = ms => new Promise(res => setTimeout(res, ms));

async function safeGoto(page, url) {
    try {
        await page.goto(url, { waitUntil: 'load', timeout: 60000 });
    } catch (e) {
        if (e.message.includes('net::ERR_ABORTED') || e.message.includes('Timeout') || e.message.includes('timeout')) {
            console.log(`Navigation to ${url} timed out or aborted (${e.message}), but proceeding...`);
        } else {
            throw e;
        }
    }
}

async function handleLogin(page) {
    console.log(`Navigating to login page: ${LOGIN_URL}`);
    await safeGoto(page, LOGIN_URL);
    await delay(3000);

    console.log("Locating credentials input fields...");
    const usernameField = await page.waitForSelector("xpath///input[@placeholder='Please enter your user name/email/mobile number']", { timeout: 20000 });
    const passwordField = await page.waitForSelector("xpath///input[@placeholder='Enter Password']", { timeout: 20000 });

    console.log("Entering credentials...");
    await usernameField.click({ clickCount: 3 });
    await usernameField.type(USERNAME);
    await page.evaluate(el => el.dispatchEvent(new Event('input', { bubbles: true })), usernameField);
    await delay(500);

    await passwordField.click({ clickCount: 3 });
    await passwordField.type(PASSWORD);
    await page.evaluate(el => el.dispatchEvent(new Event('input', { bubbles: true })), passwordField);
    await delay(500);

    console.log("Checking Privacy Policy agreement checkbox...");
    try {
        const checkbox = await page.waitForSelector("xpath///span[contains(text(), 'Privacy Policy') or contains(text(), 'agree to')]/ancestor::label//input[@type='checkbox']", { timeout: 5000 });
        const isChecked = await page.evaluate(el => el.checked, checkbox);
        if (!isChecked) {
            console.log("Agreement checkbox not selected, clicking it...");
            await page.evaluate(el => el.click(), checkbox);
            await delay(500);
        } else {
            console.log("Agreement checkbox is already selected.");
        }
    } catch (e) {
        console.log(`Warning/Note while checking agreement checkbox: ${e.message}`);
    }

    console.log("Clicking Login button...");
    const loginBtn = await page.waitForSelector("xpath///*[contains(@class, 'submit-button') or contains(@class, 'login-btn')]", { visible: true, timeout: 15000 });
    await loginBtn.click();

    console.log("Waiting for page redirect and dashboard load...");
    await page.waitForFunction(() => window.location.href.includes('/green/#/'), { timeout: 30000 });
    console.log("Login successful!");
    await delay(3000);
}

async function goToPlantsPage(page) {
    console.log("Navigating to Plants section...");
    const plantsTab = await page.waitForSelector("xpath///span[text()='Plants']", { timeout: 40000 });
    await plantsTab.click();

    console.log("Waiting for Plants table data to load...");
    try {
        await page.waitForFunction(() => !document.documentElement.classList.contains('nprogress-busy'), { timeout: 40000 });
        await page.waitForSelector('.arco-table-td', { timeout: 20000 });
        await delay(3000); // Extra render buffer
    } catch (e) {
        console.log(`Warning: Timed out waiting for table to load: ${e.message}`);
    }
}

async function scrapePlantsTable(page) {
    console.log("Scraping Plants overview table...");
    return await page.evaluate(() => {
        try {
            const headers = Array.from(document.querySelectorAll('.arco-table-th')).map(th => th.innerText.trim().replace(/\n/g, ' '));
            const rows = document.querySelectorAll('.arco-table-tr');
            const results = [];
            rows.forEach(row => {
                const cells = Array.from(row.querySelectorAll('.arco-table-td'));
                if (cells.length > 0) {
                    const rowData = {};
                    cells.forEach((cell, idx) => {
                        const header = headers[idx] || `Column_${idx}`;
                        rowData[header] = cell.innerText.trim().replace(/\n/g, ' ');
                    });
                    results.push(rowData);
                }
            });
            return results;
        } catch (e) {
            return [];
        }
    });
}

async function scrapePlantDetails(page) {
    console.log("Scraping plant detailed metrics from dashboard...");
    return await page.evaluate(() => {
        try {
            const metrics = {};
            const elements = Array.from(document.querySelectorAll('*'));

            const findValue = (label) => {
                const candidates = elements.filter(el => {
                    const txt = (el.innerText || "").trim().toLowerCase();
                    return txt.includes(label.toLowerCase());
                });

                if (candidates.length === 0) return "";

                // Sort by children count ascending to get the deepest element containing the label first
                candidates.sort((a, b) => a.getElementsByTagName('*').length - b.getElementsByTagName('*').length);

                for (const labelEl of candidates) {
                    let current = labelEl;
                    for (let i = 0; i < 4; i++) {
                        if (!current) break;
                        const text = current.innerText;
                        const match = text.match(/(\d+(?:\.\d+)?)\s*(kWh|kWp|kW|%)/i);
                        if (match) {
                            return match[0];
                        }
                        current = current.parentElement;
                    }
                }
                return "";
            };

            metrics['Daily solar'] = findValue('Daily solar');
            metrics['Daily consumption'] = findValue('Daily consumption');
            metrics['PV Capacity'] = findValue('PV Capacity');
            metrics['PV Power'] = findValue('PV Power');
            metrics['Imported energy'] = findValue('Imported energy today') || findValue('Imported energy');

            return metrics;
        } catch (e) {
            return {};
        }
    });
}

function saveToJson(newRows) {
    console.log(`Saving data to JSON file: ${JSON_FILE}`);
    let existingRows = [];
    let plantsData = [];

    if (fs.existsSync(JSON_FILE)) {
        try {
            const fileData = JSON.parse(fs.readFileSync(JSON_FILE, 'utf8'));
            existingRows = fileData.telemetry || [];
        } catch (e) {
            console.error(`Warning: Could not read existing Telemetry from JSON (${e.message}).`);
        }
    }

    const combined = [...existingRows, ...newRows];

    combined.sort((a, b) => {
        const timeA = new Date(a.timestamp).getTime();
        const timeB = new Date(b.timestamp).getTime();
        return timeB - timeA;
    });

    const seen = new Set();
    const unique = [];
    for (const record of combined) {
        const key = `${record.plant_id}|||${record.timestamp}`;
        if (!seen.has(key)) {
            seen.add(key);
            unique.push(record);
        }
    }

    unique.sort((a, b) => {
        const timeA = new Date(a.timestamp).getTime();
        const timeB = new Date(b.timestamp).getTime();
        if (timeA !== timeB) return timeA - timeB;
        return a.plant_id - b.plant_id;
    });

    unique.forEach((record, idx) => {
        record.id = idx + 1;
    });

    const colsOrder = ["id", "plant_id", "timestamp", "power", "voltage", "current", "frequency", 
                      "irradiance", "daily_generation", "total_generation", "temperature", "status", "raw_json", "created_at"];
    
    const finalRows = unique.map(r => {
        const row = {};
        for (const col of colsOrder) {
            row[col] = r[col] !== undefined ? r[col] : null;
        }
        return row;
    });

    const scrapedPlantIds = new Set(finalRows.map(r => r.plant_id));
    plantsData = Object.entries(PLANT_MAPPING)
        .map(([name, id]) => ({
            id: id,
            plant_name: name.toUpperCase()
        }))
        .filter(p => scrapedPlantIds.has(p.id))
        .sort((a, b) => a.id - b.id);

    const exportData = {
        plants: plantsData,
        telemetry: finalRows
    };

    try {
        fs.writeFileSync(JSON_FILE, JSON.stringify(exportData, null, 2), 'utf8');
        console.log(`\nSUCCESS! Telemetry updated. Total ${finalRows.length} rows written.`);
    } catch (e) {
        console.error(`Could not write JSON file:`, e);
    }
}

async function main() {
    console.log("Initializing Puppeteer Chrome browser...");
    const browser = await puppeteer.launch({
        headless: HEADLESS,
        args: [
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage',
            '--disable-gpu'
        ]
    });

    const [page] = await browser.pages();
    await page.setViewport({ width: 1920, height: 1080 });
    await page.setUserAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36");

    const allRows = [];

    try {
        // 1. Log in
        await handleLogin(page);

        // 2. Navigate to Plants page
        await goToPlantsPage(page);

        // 3. Scrape main plants list table
        const plants = await scrapePlantsTable(page);
        if (!plants || plants.length === 0) {
            throw new Error("No plants data found in the overview table.");
        }
        console.log(`Found ${plants.length} plants in the table.`);

        // Format scrape time
        const now = new Date();
        const year = now.getFullYear();
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const date = String(now.getDate()).padStart(2, '0');
        const hours = String(now.getHours()).padStart(2, '0');
        const minutes = String(now.getMinutes()).padStart(2, '0');
        const seconds = String(now.getSeconds()).padStart(2, '0');
        const scrapeTime = `${year}-${month}-${date} ${hours}:${minutes}:${seconds}`;

        // 4. Loop through each plant to collect detailed page metrics
        for (let i = 0; i < plants.length; i++) {
            const plant = plants[i];
            let plantName = (plant["Plant Name"] || "").trim();
            if (!plantName) {
                plantName = (plant[Object.keys(plant)[0]] || "").trim();
            }

            console.log(`\n[${i + 1}/${plants.length}] Processing plant: '${plantName}'`);
            console.log(`Navigating to plant details for '${plantName}'...`);

            // Listen for new tab open
            const newTargetPromise = new Promise(resolve => browser.once('targetcreated', target => resolve(target.page())));

            const plantLink = await page.waitForSelector(`xpath///table//tr//td[contains(., "${plantName}")]//*[text()="${plantName}"] | //*[text()="${plantName}"]`, { timeout: 15000 });
            await page.evaluate(el => el.click(), plantLink);

            // Wait for new tab or check if it loaded in the same tab
            let detailPage = page;
            let openedInNewTab = false;
            const newPage = await Promise.race([
                newTargetPromise,
                delay(3000).then(() => null)
            ]);

            if (newPage) {
                detailPage = newPage;
                openedInNewTab = true;
                console.log("Opened in a new tab.");
            } else {
                console.log("Did not open in a new tab (same tab navigation).");
            }

            console.log("Waiting for plant details page to finish loading...");
            try {
                // Wait for loading indicator
                await detailPage.waitForFunction(() => !document.documentElement.classList.contains('nprogress-busy'), { timeout: 25000 });
                // Wait for 'PV Power' label to be present
                await detailPage.waitForSelector("xpath///*[text()='PV Power' or contains(text(), 'PV Power')]", { timeout: 15000 });
                await delay(4000); // Sleep 4 seconds to ensure dynamic numeric values populate
            } catch (e) {
                console.log(`Warning: Timed out waiting for details page elements: ${e.message}`);
            }

            // Scrape detailed metrics
            const details = await scrapePlantDetails(detailPage);

            // Close detail tab or navigate back
            if (openedInNewTab) {
                await detailPage.close();
            } else {
                console.log("Navigating back to Plants overview page...");
                await safeGoto(page, "https://global.solaxcloud.com/green/#/plant/index");
                await delay(4000);
            }

            // Compile merged row data
            const rowData = {
                "plant_id": parseInt(getPlantId(plantName)),
                "timestamp": cleanTimestamp(scrapeTime),
                "power": parseValueWithUnit(details["PV Power"] || plant["PV Capacity(kWp)"] || plant["PV Capacity"]),
                "voltage": null,
                "current": null,
                "frequency": null,
                "irradiance": null,
                "daily_generation": parseValueWithUnit(details["Daily solar"] || plant["Daily Yield(kWh)"] || plant["Daily Yield"]),
                "total_generation": null,
                "temperature": null,
                "status": plant["Plant Status"] || "Normal",
                "raw_json": JSON.stringify({...plant, ...details}),
                "created_at": new Date().toISOString().replace('T', ' ').substring(0, 19)
            };
            allRows.push(rowData);
        }

        // 5. Export compiled dataset to JSON
        if (allRows.length > 0) {
            saveToJson(allRows);
        }

    } catch (e) {
        console.error(`\nAn error occurred during scraping:`, e);
        try {
            console.log(`Current URL at error: ${page.url()}`);
            const screenshotPath = path.join(__dirname, "error_screenshot.png");
            await page.screenshot({ path: screenshotPath });
            console.log(`Saved error screenshot to: ${screenshotPath}`);
        } catch (se) {
            console.log(`Could not save screenshot: ${se.message}`);
        }
    } finally {
        console.log("\nClosing Chrome browser...");
        await browser.close();
    }
}

main();
